Show tutorial at start.

amanagr opened the pull request:
@neiljp if you think this PR is easier to review, then I will push the changes in the other PR. I did a complete change of commits. Let me know your thoughts on this.

amanagr committed:
TUTORIAL: Add tutorial to be shown at startup.

amanagr committed:
helper: Add method to change/set a setting in zuliprc.

This will be helpful in updating settings while zt is running.

amanagr committed:
views: Add LoadingView which will be used to display tutorial.

amanagr committed:
core: Extract loop initialization into a separate function.

This will allow up to set custom loop properties without
running it.

amanagr committed:
run: Stop showing running info at startup.

Running info will now be shown via LoadingView.

neiljp reviewed (commented):
@amanagr I do like the (many) features in this and the other PR and appreciate you breaking this down differently, and it's definitely clearer to some degree. I've left some comments inline regarding specific code in this PR.

I think it would be much easier to follow if you split this up into smaller feature change PRs; I've not analyzed the commits fully, but this could include:
* Add tutorial in minimal way, *after* current loading screen, always loading it, noting how to disable
* Migrate the loading part 'into' the tutorial (ie. the LoadingView!)
* Add support to save zuliprc, possibly refactoring others with it into 'zuliprc-handling' module
* Enable toggling tutorial from loading screen
* Enable toggling other values from loading screen

Each of those should be easy to separate and understand - it's not terrible to always load the tutorial at first, for example, particularly if we're working on the next step in the next iteration.

The order is not critical, within limits, eg. you could make the LoadingView and place the actual loading into it *first*, and then add the tutorial.

I know that we didn't get feedback/reviews on this early, which may explain why it grew, but you have good code now, so working on splitting it up should be easier?

amanagr commented on zulipterminal/core.py:
```diff
@@ -62,6 +64,15 @@ def spinning_cursor() -> Any:
         sys.stdout.write('\n')
         self.capture_stdout()
 
+    def update_zuliprc(self, setting: str, value: str) -> None:
```
Okay, I will move it into the Controller.

amanagr commented on zulipterminal/core.py:
```diff
@@ -309,7 +311,7 @@ def exit_handler(self, signum: int, frame: Any) -> None:
         sys.exit(0)
 
-    def main(self) -> None:
+    def initialize_loop(self) -> None:
```
I had removed it in the later commits in the last PR too. I think you mentioned that you didn't review the last PR fully.

amanagr commented:
@neiljp do the splits look fine? I will proceed to adding tests then...

amanagr committed:
core: Show LoadingView at start instead of printing loading status.

amanagr committed:
tutorial: Wait after loading for user to read the tutorial.

amanagr commented:
![zt_tutorial](https://user-images.githubusercontent.com/25124304/76143014-3beef400-6099-11ea-9bec-045ab6202dd7.gif)
Apart from the new isort rule errors, everything works fine. @neiljp @sumanthvrao Let me know your thoughts on this.

kaustubh-nair reviewed (commented):
@amanagr Wow, this PR has a lot of features! :smile: 

Though the commits are well split, I have a few ideas to make minor changes - so that they flow better, and become easier to review + merge:
* Commit 1 can be merged as it is, since it is a refactor, and doesn't rely on future commits.
* Add `LoadingView` (Commit 2 + 3)
* Commit 5 can be modified to display `LoadingView` after loading status / model initialization.
* Commit 7 can be added here, and modified to only deal with the `enter` key - for waiting.

I imagine, the above commits should summarize the core facets of this PR. 

The other commits can be added as follow-ups/subsequent commits.
* Remove loading status ( Commit 4 )
* Change settings in zuliprc ( Commit 6 ) 
* Allow changing settings on the fly ( Commit 8 )
* Skip tutorial during future startups with `h` ( Partially in commit 7 )
* Maybe it would be useful to even have a flag to display the tutorial again during startup, instead of changing it in the config file?

amanagr commented on zulipterminal/config/tutorial.py:
```diff
@@ -0,0 +1,18 @@
+# Tutorial to display at start
+TUTORIAL = [
+    ('active', 'Welcome to Zulip!\n'),
+    '\nPlease go through this totorial to understand the \n'
```
The tutorial file is not final yet. I  have a lot of things to add to it.

amanagr commented:
Thanks @kaustubh-nair for your review, let me get back to you with what I think of them.

amanagr committed:
views: Allow changing settings from the LoadingView on the fly.

Add various keys to LoadingView which allow changing config of
themes, autohide and notify directly.

neiljp commented:
@amanagr You have one commit that removes a lot, and then another commit that adds a lot of features; could you combine them into a 'transition' commit? Otherwise, doesn't this lead to a broken build between commits? Even if that isn't considered as big a deal, it's more difficult to understand what the change is, as it's over two commits?

Also, the tests you remove in one commit - isn't that functionality that we still want to include (failure to load) ?
